src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Extensibility/ProjectExportProvider.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Composition;
+using Microsoft.VisualStudio.ProjectSystem.VS.Utilities;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Extensibility
+{
+    /// <summary>
+    /// MEF component which has methods for consumers to get to project specific MEF exports
+    /// </summary>
+    [Export(typeof(IProjectExportProvider))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    internal class ProjectExportProvider : IProjectExportProvider
+    {
+        /// <summary>
+        /// This function requires going to the UI thread to map the hierarchy to the correct project
+        /// instance. 
+        /// </summary>
+        public T GetExport<T>(IVsHierarchy projectHierarchy) where T : class
+        {
+            T exportedValue = null;
+
+            if (projectHierarchy == null)
+            {
+                return exportedValue;
+            }
+            try
+            {
+                // The following code must be run on the UI thread
+                Shell.ThreadHelper.JoinableTaskFactory.Run(async () =>
+                {
+                    await Shell.ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                    UnconfiguredProject unconfigProject = projectHierarchy.GetUnconfiguredProject();
+                    if (unconfigProject != null && unconfigProject.Capabilities.Contains(ProjectCapability.CSharpOrVisualBasic))
+                    {
+                        exportedValue = unconfigProject.Services.ExportProvider.GetExportedValue<T>();
+                    }
+                });
+            }
+            catch (Exception)
```

Comment:
Can we remove the catch here - it hides bugs.

Should replace this with a call to GetExportedValueOrDefault<T>() as someone may be marked with the CSharp capability but not have our components.